src/System.Text.Json/src/System/Text/Json/Serialization/JsonPropertyInfoCommon.cs

```diff
@@ -121,37 +121,45 @@ public override IEnumerable CreateDerivedEnumerableInstance(JsonPropertyInfo col
         {
             object instance = collectionPropertyInfo.DeclaredTypeClassInfo.CreateObject();
 
-            if (instance is IList instanceOfIList)
+            try
             {
-                foreach (object item in sourceList)
+                if (instance is IList instanceOfIList && !instanceOfIList.IsReadOnly)
                 {
-                    instanceOfIList.Add(item);
+                    foreach (object item in sourceList)
+                    {
+                        instanceOfIList.Add(item);
```

Comment:
Would it be worth adding a special-case here for `List<T>` in order to use `List<T>.AddRange()`?

Reply:
Maybe, but probably not for this PR?